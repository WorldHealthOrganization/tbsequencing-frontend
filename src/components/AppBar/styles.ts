import { css } from '@emotion/react';
import appColors from '../../styles/colors';
import { fontSizes } from '../typography/fontSizes';

const PADDING_SIZE = 54;

export const muiAppBarStyles = css({
  paddingRight: PADDING_SIZE,
  paddingLeft: PADDING_SIZE,
  backgroundColor: appColors.neutral.white,
});

export const appBarWrapperStyles = css({
  display: 'flex',
  justifyContent: 'space-between',
  position: 'relative',
  alignItems: 'stretch',
});

export const bottomLineStyles = css({
  height: 4,
  backgroundColor: appColors.tints.whoBlue.lightest,
  position: 'absolute',
  bottom: 0,
  left: 0,
  right: 0,
});

export const loaderContainer = css({
  position: 'relative',
  visibility: 'hidden',
});

export const logo = css({
  display: 'flex',
  alignItems: 'center',
  color: appColors.primary.navy,
  textDecoration: 'none',
});

export const dropdownContainer = css({
  width: '176px',
});

export const dropdownButtonStyles = css({
  fontSize: fontSizes.h2,
  paddingTop: '26px',
  paddingBottom: '26px',
  borderRadius: 0,
  color: appColors.primary.navy,
  textTransform: 'capitalize',
  borderBottom: '4px solid transparent',
  width: '160px',
  zIndex: 1,
  paddingRight: '32px',
  ':hover': {
    backgroundColor: 'transparent',
  },
  '& .MuiButton-endIcon': {
    alignSelf: 'flex-start',
  },
  '& .MuiPaper-root': {
    width: '160px',
  },
});

export const dropdownMenu = css({
  width: '170px',
});

export const dropdownActiveStyles = css({
  borderBottom: `4px solid ${appColors.secondary.whoBlue}`,
});

export const dropdownItemStyles = css({
  width: '106px',
  padding: '20px 0',
  margin: '0 16px',
  lineHeight: '28px',
  justifyContent: 'center',
  fontSize: fontSizes.h2,
  color: appColors.primary.navy,
  fontWeight: 500,
  borderBottom: `2px solid ${appColors.tints.whoBlue.lightest}`,
  '&:last-child': {
    borderBottomColor: 'transparent',
  },
  ':hover': {
    backgroundColor: 'transparent',
  },
});

export const menuWrapper = css({
  display: 'flex',
});

export const arrowIcon = css({
  color: appColors.secondary.whoBlue,
  fontSize: '40px!important',
  position: 'absolute',
  marginLeft: '8px',
  right: 0,
});
