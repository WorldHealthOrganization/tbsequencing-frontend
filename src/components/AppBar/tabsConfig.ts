import { appRoutes } from '../../navigation/appRoutes';

export const dropdownConfig = [
  { ...appRoutes.about, label: 'About' },
  { ...appRoutes.overview, label: 'Overview' },
  { ...appRoutes.drugs, label: 'Drugs' },
  { ...appRoutes.genes, label: 'Genes' },
];

export const tabsConfig = [
  { ...appRoutes.mutations, label: 'Mutation' },
  { ...appRoutes.genotypeResistance, label: 'Genotype Resistance' },
  { ...appRoutes.dataSubmission, label: 'Data submission' },
  { ...appRoutes.download, label: 'Download' },
];

export const userActiveTabsConfig: ITabsConfig = [
  {
    ...appRoutes.registration,
    label: 'Registration',
  },
  {
    ...appRoutes.dataSubmission,
    label: 'Data submission',
  },

];

export type ITabsConfig = typeof tabsConfig;
export type IDropdownConfig = typeof dropdownConfig;
