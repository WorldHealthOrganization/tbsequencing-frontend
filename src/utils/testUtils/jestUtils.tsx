import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
