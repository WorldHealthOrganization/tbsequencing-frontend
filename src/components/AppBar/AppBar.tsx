/** @jsxImportSource @emotion/react */
import React, { useEffect, useState } from 'react';
import {
  AppBar as MUIAppBar,
  Link,
  Button,
  Menu,
  MenuItem,
} from '@mui/material';
import { useLocation, useNavigate } from 'react-router-dom';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import { useMsal } from '@azure/msal-react';
import { InteractionRequiredAuthError } from '@azure/msal-browser';
import { toast } from 'react-toastify';
import * as styles from './styles';
import { NavTabs } from './NavTabs/NavTabs';

import AuthButton from '../AuthButton';
import { tabsConfig, dropdownConfig } from './tabsConfig';
import { useIsLoggedIn } from '../../features/auth/hooks/useIsLoggedIn';
import {
  getActiveTab,
  tabRoutesMatcher,
  getActiveDropdown,
} from './NavTabs/utils/tabRoutesMatcher';
import { useDrawerApi } from '../../features/drawer/hooks/useDrawerApi';
import ActivityIndicator from '../ActivityIndicator';
import H1 from '../typography/H1';
import { loginRequest } from '../../authConfig';
import { setTokens } from '../../features/auth/storeToken';
import { setCredentialsSSO, setUserSSO } from '../../features/auth/authSlice';
import { useAppDispatch } from '../../app/hooks';

export const AppBar = () => {
  const { instance, inProgress, accounts } = useMsal();
  const isLoggedIn = useIsLoggedIn();
  const navigate = useNavigate();
  const location = useLocation();
  const dispatch = useAppDispatch();
  const tabsRouteMatched = tabRoutesMatcher(location, tabsConfig);
  const dropdownButtonActive = tabRoutesMatcher(location, dropdownConfig);
  const { setDrawerContent, toggleDrawer } = useDrawerApi();

  const activeAccount = instance.getAllAccounts();

  const environmentName = process.env.REACT_APP_ENV_HEADER_NAME || 'TB Sequencing';

  const authButtonLabel = activeAccount[0]?.name
    ? activeAccount[0].name
    : 'Sign in';

  const initialTab = getActiveTab(location, tabsConfig);
  const initialDropdown = getActiveDropdown(
    location,
    dropdownConfig,
    'Overview',
  );

  const [activeTab, setActiveTab] = useState<number>(initialTab);
  const [selectedMenuItem, setSelectedMenuItem] = useState<string>(initialDropdown);
  const [isDropdownActive, setIsDropdownActive] = useState<boolean>(false);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);

  useEffect(() => {
    const userName = activeAccount ? activeAccount[0]?.username : '';
    function getTokenRedirect(request: any) {
      /**
       * See here for more info on account retrieval:
       * https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-common/docs/Accounts.md
       */
      request.account = instance.getAccountByUsername(userName);

      return instance.acquireTokenSilent(request).catch((error) => {
        toast.info(
          'silent token acquisition fails. acquiring token using redirect',
        );
        if (error instanceof InteractionRequiredAuthError) {
          // fallback to interaction when silent call fails
          return instance.acquireTokenRedirect(request);
        }
        return null;
      });
    }

    setActiveTab(getActiveTab(location, tabsConfig));
    setIsDropdownActive(dropdownButtonActive);
    setSelectedMenuItem(
      getActiveDropdown(location, dropdownConfig, selectedMenuItem),
    );
    if (accounts.length) {
      getTokenRedirect(loginRequest)
        .then((response) => {
          if (response?.accessToken) {
            setTokens(response?.accessToken, null);
            dispatch(setCredentialsSSO(response?.accessToken));
            dispatch(setUserSSO(activeAccount[0]));
          }
        })
        .catch((error) => {
          toast.error(error);
        });
    }
  }, [location,
    dropdownButtonActive,
    selectedMenuItem,
    activeAccount,
    accounts,
    dispatch,
    instance]);

  const handleTabChange = (event: React.SyntheticEvent, value: number) => {
    const selectedRoute = tabsConfig[value];
    setIsDropdownActive(false);

    setActiveTab(value);
    if ((!isLoggedIn || !activeAccount[0]) && selectedRoute.label === 'Data submission') {
      setDrawerContent('anonymous');
      toggleDrawer();
    } else {
      navigate(selectedRoute.path);
    }
  };

  const anonymousOnAuthButtonClick = () => {
    setDrawerContent('anonymous');
    toggleDrawer();
  };

  const loggedInButtonClick = () => {
    setDrawerContent('userMenu');
    toggleDrawer();
  };

  const onAuthButtonClick = isLoggedIn && activeAccount[0]
    ? loggedInButtonClick
    : anonymousOnAuthButtonClick;

  const handleOpenDropdown = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleCloseDropdown = () => {
    setAnchorEl(null);
  };

  const handleSelectMenuItem = (menuItem: string, index: number) => {
    setSelectedMenuItem(menuItem);
    handleCloseDropdown();
    const selectedRoute = dropdownConfig[index];
    setIsDropdownActive(true);

    setActiveTab(index);
    navigate(selectedRoute.path);
  };

  return (
    <MUIAppBar
      elevation={0}
      position="sticky"
      css={styles.muiAppBarStyles}
      color="transparent"
    >
      <div css={styles.appBarWrapperStyles}>
        <Link css={styles.logo} href="/overview">
          <H1>{environmentName}</H1>
        </Link>
        <div css={styles.menuWrapper}>
          <div css={styles.dropdownContainer}>
            <Button
              id="dropdown-button"
              aria-controls={open ? 'dropdown-menu' : undefined}
              aria-haspopup="true"
              aria-expanded={open ? 'true' : undefined}
              onClick={handleOpenDropdown}
              css={[
                styles.dropdownButtonStyles,
                isDropdownActive && styles.dropdownActiveStyles,
              ]}
              endIcon={<KeyboardArrowDownIcon css={styles.arrowIcon} />}
            >
              {selectedMenuItem}
            </Button>
            <Menu
              id="dropdown-menu"
              anchorEl={anchorEl}
              open={open}
              onClose={handleCloseDropdown}
              MenuListProps={{
                'aria-labelledby': 'dropdown-button',
              }}
              css={styles.dropdownMenu}
            >
              {dropdownConfig.map((dropItem, index) => (
                <MenuItem
                  key={dropItem.label}
                  onClick={() => handleSelectMenuItem(dropItem.label, index)}
                  css={styles.dropdownItemStyles}
                >
                  {dropItem.label}
                </MenuItem>
              ))}
            </Menu>
          </div>

          <NavTabs
            routeMatched={tabsRouteMatched}
            tabsConfig={tabsConfig}
            onTabChange={handleTabChange}
            activeTab={activeTab}
          />
        </div>
        {!inProgress ? (
          <div css={styles.loaderContainer}>
            <ActivityIndicator centered />
          </div>
        ) : (
          <AuthButton
            onClick={onAuthButtonClick}
            label={authButtonLabel}
            isActive={false}
            isAppBar
          />
        )}
        <div css={styles.bottomLineStyles} />
      </div>
    </MUIAppBar>
  );
};

export default AppBar;
