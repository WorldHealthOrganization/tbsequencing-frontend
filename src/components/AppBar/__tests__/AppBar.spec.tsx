import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, getTreeJSONWithRedux } from '../../../utils/testUtils/jestUtils';
import AppBar from '../AppBar';

describe('AppBar component test', () => {
  test('renders AppBar component without error', () => {
    render(
      <BrowserRouter>
        <AppBar />
      </BrowserRouter>
      ,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <BrowserRouter>
        <AppBar />
      </BrowserRouter>,
    );

    expect(tree).toMatchSnapshot();
  });
});
