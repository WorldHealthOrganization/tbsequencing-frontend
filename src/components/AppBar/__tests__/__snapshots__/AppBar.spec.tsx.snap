// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar component test Should render correctly 1`] = `
<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionSticky css-ec2uxu-MuiPaper-root-MuiAppBar-root-AppBar"
>
  <div
    className="css-1n2ro0c-AppBar"
  >
    <a
      className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1wib7k3-MuiTypography-root-MuiLink-root-AppBar"
      href="/overview"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1i0qxbv-MuiTypography-root-BaseText"
      >
        TB Sequencing
      </p>
    </a>
    <div
      className="css-1wfclej-AppBar"
    >
      <div
        className="css-1xnrgq5-AppBar"
      >
        <button
          aria-haspopup="true"
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-c2ys9b-MuiButtonBase-root-MuiButton-root-AppBar"
          disabled={false}
          id="dropdown-button"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Overview
          <span
            className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1qnpn9f-MuiSvgIcon-root-AppBar"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="MuiTabs-root css-1oujr26-MuiTabs-root"
      >
        <div
          className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          onScroll={[Function]}
          style={
            Object {
              "marginBottom": -0,
              "overflow": "hidden",
            }
          }
        >
          <div
            aria-orientation={null}
            className="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            onKeyDown={[Function]}
            role="tablist"
          >
            <button
              aria-selected={false}
              className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-dpb51x-MuiButtonBase-root-MuiTab-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex={0}
              type="button"
            >
              Mutation
            </button>
            <button
              aria-selected={false}
              className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-dpb51x-MuiButtonBase-root-MuiTab-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex={-1}
              type="button"
            >
              BioSample search
            </button>
            <button
              aria-selected={false}
              className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-dpb51x-MuiButtonBase-root-MuiTab-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex={-1}
              type="button"
            >
              Data submission
            </button>
            <button
              aria-selected={false}
              className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-dpb51x-MuiButtonBase-root-MuiTab-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex={-1}
              type="button"
            >
              Download
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1y2xnzi-AuthButton"
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1onmchy-MuiButtonBase-root-MuiButton-root-AuthButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <span
            className="css-9rttau-AuthButton"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1083c97-MuiSvgIcon-root"
              data-testid="AccountBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 5v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2H5c-1.11 0-2 .9-2 2zm12 4c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3zm-9 8c0-2 4-3.1 6-3.1s6 1.1 6 3.1v1H6v-1z"
              />
            </svg>
          </span>
        </span>
        Sign in
      </button>
    </div>
    <div
      className="css-toebm4-AppBar"
    />
  </div>
</header>
`;
